File: scripts/optimize_assets.py
import os
import glob
from PIL import Image

# Configuration
ASSETS_DIR = "public/assets"
SQUADRA_DIR = os.path.join(ASSETS_DIR, "squadra")
ALBUMS_DIR = os.path.join(ASSETS_DIR, "albums")

# Files to keep (root assets only)
# These are the DSC files referenced in the code + logo
KEEP_FILES = {
    "DSC08495.jpg",
    "DSC08524.jpg",
    "DSC08437.jpg",
    "DSC08451.jpg",
    "DSC08466.jpg",
    "logo.png"
}

def convert_to_webp(filepath):
    try:
        filename = os.path.basename(filepath)
        name, ext = os.path.splitext(filename)
        if ext.lower() == ".webp":
            return filepath # Already webp

        new_filepath = os.path.join(os.path.dirname(filepath), name + ".webp")
        
        print(f"Converting {filename} to WebP...")
        with Image.open(filepath) as img:
            img.save(new_filepath, "WEBP", quality=85)
        
        # Verify creation before deletion
        if os.path.exists(new_filepath):
            os.remove(filepath)
            print(f"Deleted original {filename}")
            return new_filepath
        else:
            print(f"Error: Failed to create {new_filepath}")
            return filepath
            
    except Exception as e:
        print(f"Failed to convert {filepath}: {e}")
        return filepath

def process_root_assets():
    print("Processing root assets...")
    # List all files in assets dir
    for filepath in glob.glob(os.path.join(ASSETS_DIR, "*")):
        if os.path.isdir(filepath):
            continue
            
        filename = os.path.basename(filepath)
        
        # Check if it's a DSC image
        if filename.startswith("DSC") and filename.lower().endswith(".jpg"):
            if filename in KEEP_FILES:
                convert_to_webp(filepath)
            else:
                print(f"Deleting unused asset: {filename}")
                os.remove(filepath)
        
        # Check for galacticos*.webp
        elif filename.startswith("galacticos") and filename.lower().endswith(".webp"):
            print(f"Deleting unused asset: {filename}")
            os.remove(filepath)
            
        # Check for logo
        elif filename == "logo.png":
            convert_to_webp(filepath)

def process_directory(directory):
    print(f"Processing directory: {directory}")
    for filepath in glob.glob(os.path.join(directory, "*")):
        if os.path.isdir(filepath):
            continue
        
        filename = os.path.basename(filepath)
        if filename.lower().endswith(('.jpg', '.jpeg', '.png')):
             convert_to_webp(filepath)

if __name__ == "__main__":
    if not os.path.exists(ASSETS_DIR):
        print(f"Directory not found: {ASSETS_DIR}")
        exit(1)
        
    process_root_assets()
    process_directory(SQUADRA_DIR)
    process_directory(ALBUMS_DIR)
    print("Optimization complete.")

File: scripts/optimize-assets.js
const fs = require('fs');
const path = require('path');
const sharp = require('sharp');

const ASSETS_DIR = path.join(process.cwd(), 'public', 'assets');
const SQUADRA_DIR = path.join(ASSETS_DIR, 'squadra');
const ALBUMS_DIR = path.join(ASSETS_DIR, 'albums');

// Files to keep (root assets only)
// Based on code analysis + logo
const KEEP_FILES = new Set([
    'DSC08495.jpg', 'DSC08495.webp',
    'DSC08524.jpg', 'DSC08524.webp',
    'DSC08437.jpg', 'DSC08437.webp',
    'DSC08451.jpg', 'DSC08451.webp',
    'DSC08466.jpg', 'DSC08466.webp',
    'logo.png', 'logo.webp'
]);

async function optimizeImage(filepath) {
    const filename = path.basename(filepath);
    const ext = path.extname(filename).toLowerCase();
    const basename = path.basename(filename, ext);
    
    // Target is always webp
    const newFilepath = path.join(path.dirname(filepath), `${basename}.webp`);

    console.log(`Optimizing ${filename}...`);
    try {
        // Load image
        let pipeline = sharp(filepath);

        // Get metadata to check dimensions
        const metadata = await pipeline.metadata();
        
        // Resize if width > 1920
        if (metadata.width > 1920) {
            console.log(`  -> Resizing from ${metadata.width}px to 1920px width`);
            pipeline = pipeline.resize({ width: 1920 });
        }

        // Save as WebP with lower quality for size reduction
        // Using a temporary buffer to handling overwriting the same file if input is webp
        const buffer = await pipeline
            .webp({ quality: 75 }) 
            .toBuffer();

        fs.writeFileSync(newFilepath, buffer);
        
        // If input was NOT webp, delete the original
        if (ext !== '.webp' && fs.existsSync(newFilepath)) {
            console.log(`  -> Deleted original ${filename}`);
            fs.unlinkSync(filepath);
        } else {
             console.log(`  -> Re-saved ${newFilepath}`);
        }
        
    } catch (err) {
        console.error(`Failed to optimize ${filename}:`, err);
    }
}

async function processRootAssets() {
    console.log('Processing root assets...');
    const files = fs.readdirSync(ASSETS_DIR);
    
    for (const file of files) {
        if (fs.statSync(path.join(ASSETS_DIR, file)).isDirectory()) continue;
        
        const ext = path.extname(file).toLowerCase();
        
        // Process ONLY the keep files (both jpg and existing webp)
        if (KEEP_FILES.has(file)) {
             await optimizeImage(path.join(ASSETS_DIR, file));
        } 
        // We already deleted unused stuff in previous step, but let's be safe and ignore others
    }
}

async function processDirectory(directory) {
    if (!fs.existsSync(directory)) return;
    console.log(`Processing directory: ${directory}`);
    const files = fs.readdirSync(directory);
    
    for (const file of files) {
         if (fs.statSync(path.join(directory, file)).isDirectory()) continue;
         const ext = path.extname(file).toLowerCase();
         
         // Process all images in subfolders
         if (['.jpg', '.jpeg', '.png', '.webp'].includes(ext)) {
             await optimizeImage(path.join(directory, file));
         }
    }
}

async function main() {
    await processRootAssets();
    await processDirectory(SQUADRA_DIR);
    await processDirectory(ALBUMS_DIR);
    console.log('Optimization complete.');
}

main().catch(console.error);
